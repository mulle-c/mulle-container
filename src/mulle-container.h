//
//  mulle-container.h
//  mulle-container
//
//  Created by Nat! on 02/11/15.
//  Copyright (c) 2015 Nat! - Mulle kybernetiK.
//  Copyright (c) 2015 Codeon GmbH.
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  Neither the name of Mulle kybernetiK nor the names of its contributors
//  may be used to endorse or promote products derived from this software
//  without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
//  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
//  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
//  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
//  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
//  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
//  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
//  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
//  POSSIBILITY OF SUCH DAMAGE.
//
#ifndef mulle_container__h__
#define mulle_container__h__

#define MULLE_CONTAINER_VERSION  ((2 << 20) | (0 << 8) | 0)

#include "include.h"

#include "mulle-container-operation.h"
#include "mulle-container-math.h"
#include "mulle-container-callback.h"
#include "mulle-container-callback-global.h"

#include "mulle-prime.h"
#include "mulle-hash.h"

#include "mulle--array.h"
#include "mulle-array.h"

#include "mulle--set.h"
#include "mulle-set.h"

#include "mulle--map.h"
#include "mulle-map.h"

#include "mulle-range.h"
#include "mulle--rangeset.h"

#include "mulle--pointerqueue.h"
#include "mulle-pointerarray.h"
#include "mulle--pointermap.h"
#include "mulle-pointerset.h"
#include "mulle-pointerpairarray.h"

#include "mulle-container-callback.h"
#include "mulle-container-operation.h"

#if MULLE_C11_VERSION < ((3 << 20) | (2 << 8) | 0)
# error "mulle-c11 is too old"
#endif

#if MULLE_ALLOCATOR_VERSION < ((2 << 20) | (0 << 8) | 0)
# error "mulle-allocator is too old"
#endif


#endif /* mulle_container_h */
