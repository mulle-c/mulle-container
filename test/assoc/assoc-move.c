//
//  main.c
//  test-assoc
//
//  Created by Nat! on 05.11.15.
//  Copyright (c) 2015 Nat! - Mulle kybernetiK.
//  Copyright (c) 2015 Codeon GmbH.
//  All rights reserved.
//
#ifdef NDEBUG
# undef NDEBUG
#endif
#include <assert.h>

#include <stdio.h>
#include <mulle-container/mulle-container.h>
#include <mulle-testallocator/mulle-testallocator.h>


#define N  5

static void   move_test( struct mulle_range range, intptr_t index)
{
   struct mulle_assoc                        *assoc;
   struct mulle_container_keyvaluecallback   callback;
   struct mulle_pointerpair                  pair;
   intptr_t                                  key;
   char                                      *value;
   char                                      *sep;
   size_t                              i;

   callback.keycallback   = mulle_container_keycallback_intptr;
   callback.valuecallback = mulle_container_valuecallback_copied_cstring;

   assoc = mulle_assoc_create( 0, &callback, _mulle_pointerpair_compare_pointer_key, NULL);

   mulle_assoc_add( assoc, (void *) 10, "10");
   mulle_assoc_add( assoc, (void *) 11, "11");
#if N > 2
   mulle_assoc_add( assoc, (void *) 12, "12");
#endif
#if N > 3
   mulle_assoc_add( assoc, (void *) 13, "13");
#endif
#if N > 4
   mulle_assoc_add( assoc, (void *) 14, "14");
#endif
#if N > 5
   mulle_assoc_add( assoc, (void *) 15, "15");
#endif

   _mulle_assoc_move_intptr_key_range( assoc, range, index);

   printf( "<%td-%td> -> %td : ", range.location, range.length, index);

   sep = "{";
   mulle_assoc_for( assoc, key, value)
   {
      printf( "%s %s", sep, value);
      sep = ",";
   }
   printf( " }\n");

   mulle_assoc_destroy( assoc);
}


static void   move()
{
   struct mulle_range   range;
   intptr_t             index;

   for( range.location = 10; range.location < 10 + N; range.location++)
      for( range.length = 1; range.length <= (10 + N) - range.location; range.length++)
         for( index = 10; index <= 10 + N; index++)
            move_test( range, index);
}


// the mulle_testallocator detects and aborts on leaks
static void   run_test( void (*f)( void), char *name)
{
   mulle_testallocator_reset();
#ifdef DEBUG
   fprintf( stderr, "%s\n", name);
#endif
   (f)();
   mulle_testallocator_reset();
}


int   main( int argc, char * argv[])
{
   run_test( move, "move");

   return( 0);
}
