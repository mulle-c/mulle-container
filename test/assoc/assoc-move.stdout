<10-1> -> 10 : { 10, 11, 12, 13, 14 }
<10-1> -> 11 : { 11, 10, 12, 13, 14 }
<10-1> -> 12 : { 11, 12, 10, 14, 13 }
<10-1> -> 13 : { 11, 12, 13, 10, 14 }
<10-1> -> 14 : { 11, 12, 13, 14, 10 }
<10-1> -> 15 : { 11, 12, 13, 14, 10 }
<10-2> -> 10 : { 10, 11, 12, 13, 14 }
<10-2> -> 11 : { 10, 11, 12, 13, 14 }
<10-2> -> 12 : { 12, 10, 11, 14, 13 }
<10-2> -> 13 : { 12, 13, 10, 11, 14 }
<10-2> -> 14 : { 12, 13, 14, 10, 11 }
<10-2> -> 15 : { 12, 13, 14, 10, 11 }
<10-3> -> 10 : { 10, 11, 12, 13, 14 }
<10-3> -> 11 : { 10, 11, 12, 13, 14 }
<10-3> -> 12 : { 10, 11, 12, 13, 14 }
<10-3> -> 13 : { 13, 10, 11, 12, 14 }
<10-3> -> 14 : { 13, 14, 10, 11, 12 }
<10-3> -> 15 : { 13, 14, 10, 11, 12 }
<10-4> -> 10 : { 10, 11, 12, 13, 14 }
<10-4> -> 11 : { 10, 11, 12, 13, 14 }
<10-4> -> 12 : { 10, 11, 12, 13, 14 }
<10-4> -> 13 : { 10, 11, 12, 13, 14 }
<10-4> -> 14 : { 14, 10, 11, 12, 13 }
<10-4> -> 15 : { 14, 10, 11, 12, 13 }
<10-5> -> 10 : { 10, 11, 12, 13, 14 }
<10-5> -> 11 : { 10, 11, 12, 13, 14 }
<10-5> -> 12 : { 10, 11, 12, 13, 14 }
<10-5> -> 13 : { 10, 11, 12, 13, 14 }
<10-5> -> 14 : { 10, 11, 12, 13, 14 }
<10-5> -> 15 : { 10, 11, 12, 13, 14 }
<11-1> -> 10 : { 11, 10, 12, 13, 14 }
<11-1> -> 11 : { 10, 11, 12, 13, 14 }
<11-1> -> 12 : { 10, 12, 11, 13, 14 }
<11-1> -> 13 : { 10, 12, 13, 11, 14 }
<11-1> -> 14 : { 10, 12, 13, 14, 11 }
<11-1> -> 15 : { 10, 12, 13, 14, 11 }
<11-2> -> 10 : { 11, 12, 10, 13, 14 }
<11-2> -> 11 : { 10, 11, 12, 13, 14 }
<11-2> -> 12 : { 10, 11, 12, 13, 14 }
<11-2> -> 13 : { 10, 13, 11, 12, 14 }
<11-2> -> 14 : { 10, 13, 14, 11, 12 }
<11-2> -> 15 : { 10, 13, 14, 11, 12 }
<11-3> -> 10 : { 11, 12, 13, 10, 14 }
<11-3> -> 11 : { 10, 11, 12, 13, 14 }
<11-3> -> 12 : { 10, 11, 12, 13, 14 }
<11-3> -> 13 : { 10, 11, 12, 13, 14 }
<11-3> -> 14 : { 10, 14, 11, 12, 13 }
<11-3> -> 15 : { 10, 14, 11, 12, 13 }
<11-4> -> 10 : { 11, 12, 13, 14, 10 }
<11-4> -> 11 : { 10, 11, 12, 13, 14 }
<11-4> -> 12 : { 10, 11, 12, 13, 14 }
<11-4> -> 13 : { 10, 11, 12, 13, 14 }
<11-4> -> 14 : { 10, 11, 12, 13, 14 }
<11-4> -> 15 : { 10, 11, 12, 13, 14 }
<12-1> -> 10 : { 12, 10, 11, 13, 14 }
<12-1> -> 11 : { 10, 12, 11, 13, 14 }
<12-1> -> 12 : { 10, 11, 12, 13, 14 }
<12-1> -> 13 : { 10, 11, 13, 12, 14 }
<12-1> -> 14 : { 10, 11, 13, 14, 12 }
<12-1> -> 15 : { 10, 11, 13, 14, 12 }
<12-2> -> 10 : { 12, 13, 10, 11, 14 }
<12-2> -> 11 : { 10, 12, 13, 11, 14 }
<12-2> -> 12 : { 10, 11, 12, 13, 14 }
<12-2> -> 13 : { 10, 11, 12, 13, 14 }
<12-2> -> 14 : { 10, 11, 14, 12, 13 }
<12-2> -> 15 : { 10, 11, 14, 12, 13 }
<12-3> -> 10 : { 12, 13, 14, 10, 11 }
<12-3> -> 11 : { 10, 12, 13, 14, 11 }
<12-3> -> 12 : { 10, 11, 12, 13, 14 }
<12-3> -> 13 : { 10, 11, 12, 13, 14 }
<12-3> -> 14 : { 10, 11, 12, 13, 14 }
<12-3> -> 15 : { 10, 11, 12, 13, 14 }
<13-1> -> 10 : { 13, 10, 11, 12, 14 }
<13-1> -> 11 : { 10, 13, 11, 12, 14 }
<13-1> -> 12 : { 10, 11, 13, 12, 14 }
<13-1> -> 13 : { 10, 11, 12, 13, 14 }
<13-1> -> 14 : { 10, 11, 12, 14, 13 }
<13-1> -> 15 : { 10, 11, 12, 14, 13 }
<13-2> -> 10 : { 13, 14, 10, 11, 12 }
<13-2> -> 11 : { 10, 13, 14, 11, 12 }
<13-2> -> 12 : { 10, 11, 13, 14, 12 }
<13-2> -> 13 : { 10, 11, 12, 13, 14 }
<13-2> -> 14 : { 10, 11, 12, 13, 14 }
<13-2> -> 15 : { 10, 11, 12, 13, 14 }
<14-1> -> 10 : { 14, 10, 11, 12, 13 }
<14-1> -> 11 : { 10, 14, 11, 12, 13 }
<14-1> -> 12 : { 10, 11, 14, 12, 13 }
<14-1> -> 13 : { 10, 11, 12, 14, 13 }
<14-1> -> 14 : { 10, 11, 12, 13, 14 }
<14-1> -> 15 : { 10, 11, 12, 13, 14 }
